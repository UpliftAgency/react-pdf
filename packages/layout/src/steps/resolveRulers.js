import * as R from 'ramda';

import { RULER_WIDTH } from '../constants';
import hasVerticalRuler from '../node/hasVerticalRuler';
import hasHorizontalRuler from '../node/hasHorizontalRuler';

/**
 * Adjust page size given ruler props
 *
 * @param {Object} page
 * @returns {boolean} page with size altered by ruler props
 */
const adjustPageSize = R.compose(
  R.when(
    hasVerticalRuler,
    R.evolve({
      box: { height: R.add(RULER_WIDTH) },
      children: R.map(
        R.evolve({
          box: { top: R.add(RULER_WIDTH) },
        }),
      ),
    }),
  ),
  R.when(
    hasHorizontalRuler,
    R.evolve({
      box: { width: R.add(RULER_WIDTH) },
      children: R.map(
        R.evolve({
          box: { left: R.add(RULER_WIDTH) },
        }),
      ),
    }),
  ),
);

/**
 * Adjust pages size given ruler props
 *
 * @param {Object} root
 * @returns {boolean} root with pages size altered by ruler props
 */
const resolveRulers = R.evolve({
  children: R.map(adjustPageSize),
});

export default resolveRulers;
