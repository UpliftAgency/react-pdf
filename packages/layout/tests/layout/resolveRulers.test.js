import resolveRulers from '../../src/layout/resolveRulers';

describe('layout resolveRulers', () => {
  test('Should keep page size untouched if no rulers available', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          props: {},
          box: { height: 1000, width: 500 },
        },
      ],
    };
    const result = resolveRulers(root);

    expect(result).toMatchSnapshot();
  });

  test('Should adjust only height if vertical ruler available', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          props: { verticalRuler: true },
          box: { height: 1000, width: 500 },
        },
      ],
    };
    const result = resolveRulers(root);

    expect(result).toMatchSnapshot();
  });

  test('Should adjust only width if horizontal ruler available', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          props: { horizontalRuler: true },
          box: { height: 1000, width: 500 },
        },
      ],
    };
    const result = resolveRulers(root);

    expect(result).toMatchSnapshot();
  });

  test('Should adjust height and width if both ruler available', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          props: { horizontalRuler: true, verticalRuler: true },
          box: { height: 1000, width: 500 },
        },
      ],
    };
    const result = resolveRulers(root);

    expect(result).toMatchSnapshot();
  });

  test('Should adjust height and width if ruler prop available', () => {
    const root = {
      type: 'DOCUMENT',
      children: [
        {
          type: 'PAGE',
          props: { ruler: true },
          box: { height: 1000, width: 500 },
        },
      ],
    };
    const result = resolveRulers(root);

    expect(result).toMatchSnapshot();
  });
});
