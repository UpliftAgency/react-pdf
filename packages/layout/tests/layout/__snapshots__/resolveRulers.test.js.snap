// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout resolveRulers Should adjust height and width if both ruler available 1`] = `
Object {
  "children": Array [
    Object {
      "box": Object {
        "height": 1013,
        "width": 513,
      },
      "props": Object {
        "horizontalRuler": true,
        "verticalRuler": true,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolveRulers Should adjust height and width if ruler prop available 1`] = `
Object {
  "children": Array [
    Object {
      "box": Object {
        "height": 1013,
        "width": 513,
      },
      "props": Object {
        "ruler": true,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolveRulers Should adjust only height if vertical ruler available 1`] = `
Object {
  "children": Array [
    Object {
      "box": Object {
        "height": 1013,
        "width": 500,
      },
      "props": Object {
        "verticalRuler": true,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolveRulers Should adjust only width if horizontal ruler available 1`] = `
Object {
  "children": Array [
    Object {
      "box": Object {
        "height": 1000,
        "width": 513,
      },
      "props": Object {
        "horizontalRuler": true,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolveRulers Should keep page size untouched if no rulers available 1`] = `
Object {
  "children": Array [
    Object {
      "box": Object {
        "height": 1000,
        "width": 500,
      },
      "props": Object {},
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;
